htlcswitch/switch_control.go

```diff
@@ -0,0 +1,143 @@
+package htlcswitch
+
+import (
+	"errors"
+	"sync"
+
+	"github.com/lightningnetwork/lnd/channeldb"
+	"github.com/lightningnetwork/lnd/lnwire"
+)
+
+var (
+	// ErrAlreadyPaid is used when we have already paid
+	ErrAlreadyPaid = errors.New("invoice is already paid")
+
+	// ErrPaymentInFlight returns in case if payment is already "in flight" status
+	ErrPaymentInFlight = errors.New("payment is in transition")
+
+	// ErrPaymentNotInitiated returns in case if payment wasn't initiated in switch
+	ErrPaymentNotInitiated = errors.New("payment isn't initiated")
+
+	// ErrPaymentAlreadyCompleted returns in case of attempt to complete completed payment
+	ErrPaymentAlreadyCompleted = errors.New("payment is already completed")
+)
+
+// ControlTower is a controller interface of sending HTLC messages to switch
+type ControlTower interface {
+	// CheckSend intercepts incoming message to provide checks
+	// and fail if specific message is not allowed by implementation
+	CheckSend(htlc *lnwire.UpdateAddHTLC) error
+
+	// Success marks message transition as successful
+	Success(paymentHash [32]byte) error
+
+	// Fail marks message transition as failed
+	Fail(paymentHash [32]byte) error
+}
+
+// paymentControl is implementation of ControlTower to restrict double payment sending
+type paymentControl struct {
+	db *channeldb.DB
+
+	mx sync.Mutex
+}
+
+// NewPaymentControl creates a new instance of the paymentControl.
+func NewPaymentControl(db *channeldb.DB) ControlTower {
+	return &paymentControl{
+		db: db,
+		mx: sync.Mutex{},
+	}
+}
+
+// CheckSend checks that a sending htlc wasn't triggered before for specific payment hash,
+// if so, should trigger error depends on current status
+func (p *paymentControl) CheckSend(htlc *lnwire.UpdateAddHTLC) error {
+	p.mx.Lock()
+	defer p.mx.Unlock()
+
+	// retrieve current status of payment from local database
+	paymentStatus, err := p.db.FetchPaymentStatus(htlc.PaymentHash)
+	if err != nil {
+		return err
+	}
+
+	switch paymentStatus {
+	case channeldb.StatusGrounded:
+		// it is safe to reattempt a payment if we know that we haven't
+		// left one in flight prior to restarting and switch
+		return p.db.UpdatePaymentStatus(htlc.PaymentHash, channeldb.StatusInFlight)
+
+	case channeldb.StatusInFlight:
+		// not clear if it's safe to reinitiate a payment if there
+		// is already a payment in flight, so we should withhold any
+		// additional attempts to send to that payment hash
+		return ErrPaymentInFlight
+
+	case channeldb.StatusCompleted:
+		// it has been already paid and don't want to pay again
+		return ErrAlreadyPaid
+
+	default:
+		// unknown payment status
+		return errors.New("unknown payment status")
```

Comment:
We could probably move this into `FetchPaymentStatus` in order to avoid having to check this every time.

Reply:
I've moved it to `FromBytes` method